primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
 
     /// Enforce correct `merkle_proof` for the `elem` against
     /// `expected_merkle_root`.
-    fn enforce_merkle_proof(
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// checking non-membership proof
+    fn is_non_member(
+        &mut self,
+        elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        merkle_root: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `merkle_proof` for the empty leaf `elem` against
+    /// `expected_merkle_root`.
+    fn enforce_non_membership_proof(
+        &mut self,
+        empty_elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        expected_merkle_root: Variable,
+    ) -> Result<(), CircuitError>;
+}
+
+pub(crate) trait MerkleTreeHelperGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    type MembershipProofBooleanEncoding;
+    type MembershipProofVar;
+    /// Produces an ordered list of variables based on the relative position of
+    /// a node and its siblings.
+    /// * `node` - node to be inserted in the final list.
+    /// * `sibling1` - first sibling
+    /// * `sibling2` - second sibling
+    /// * `node_is_left` - variable that is true if node is the leftmost one.
+    /// * `node_is_right` -  variable that is true if node is the rightmost one.
+    /// * `returns` - list of variables corresponding to the node and its
+    ///   siblings in the correct order.
+    fn permute(
```

Comment:
this sounds more like `sort/order/sequence`?
without reading the doc, I would think `permute(vec) -> permuted_vec`

Reply:
Yes good point. I've changed it to be more explicit, let me know if you agree with the new name and docs:
https://github.com/EspressoSystems/jellyfish/pull/172/commits/3cc4ab218c53a218d797a7a59719a34efd87137d

Reply:
yes, I like the new name, 
- meanwhile given `constrain_sibling_order()-> Result<[Variable; 3], E>`, should we rename `MerkleTreeHelperGadget` -> `3AryMerkleTreeHelperGadget`. 
- more importantly, after giving some thoughts, I question if we really need `pub(crate) trait MerkleTreeHelperGadget` 🤔 

originally, we have the helper gadget because we want to implement non-private methods on a foreign struct (i.e. `struct PlonkCircuit`). The two ways we can do it is: 1) our crate-internal helper trait; 2) just make it a function instead of a method of that struct.
After reading the usage, `constrain_membership_proof()` and `compute_merkle_root()` can be inlined as they are really called once (meaning they don't even need to be private functions at all!)
we did invoke `constrain_silbing_order()` couple of times, therefore may I suggest we have the following private stand-alone function:
`fn constrain_silbing_order<F>(circuit: &mut PlonkCircuit<F>, ...) -> ...`

I like removing unnecessary mental load on "why we need such helper gadget again" when we look at the code in the future.

Reply:
frankly the triggering cringy point is when I saw the redundant associates type `MerkleTreeHelperGadget::MembershipProofVar`, that's an indication that we might be designing our API wrong-ish.

Reply:
yes inlining `compute_merkle_root` crossed my mind too, TBH, though I was ok with leaving it in, as it was an internal impl detail. But yes, if you also see that it's a bit of a complication, I think it's best to remove the Helper gadget completely.

Reply:
Done in https://github.com/EspressoSystems/jellyfish/pull/172/files/f2d2e8e0aa915c3025ac117c040a3b3ff93e2c36..1a266b75b17276b05eca60092d78ec07b02901b5